### ISSUE TITLE
Dynamic model creation: allow type annotation to be specified along with complex fields

### ISSUE THREAD

**Issue by MatrixScenary** [samuelcolvin/pydantic#2248]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this feature/change is needed
* [x] After submitting this, I commit to one of:
    * Look through open issues and helped at least one other person
    * Hit the "watch" button on this repo to receive notifications and I commit to help at least 2 people that ask questions in the future
    * Implement a Pull Request for a confirmed bug

# Feature Request

Currently the `create_model` functions takes 3 kinds of fields: a tuple, the return of the `Field` function (`FieldInfo`) or an object (a default value). In this way there's no way to specify the type annotation when attempting to create a more complex field; which is quite a design flaw.

It looks like it could be really easy to fix. There may be an additional type which is accepted as an item of the `field_definitions` kwargs dict, and that would aggregate the `FieldInfo` and the type info. Or, there could a be special case for handling tuples, if the second item of the tuple is a `FieldInfo` object, then the first is interpreted as a type annotation.

Any thoughts?

**Comment by PrettyWood** [comment]
Hi @MatrixScenary!
Do you have an example to add to make the feature request more explicit?

**Comment by MatrixScenary** [comment]
It turns out this method works beautifully:
```python
from pydantic import create_model
field_tuple = (Optional[int], Field(None, description='Smells good.'))
model = create_model('MyModel_with_descriptions', apple_count=field_tuple)
print(model.__fields__)
```
Output:
> {'apple_count': ModelField(name='apple_count', type=Optional[int], required=False, default=None)}

Only problem is according to how it's documented, `field_tuple`'s 2nd item should be the default value to the field. It doesn't say at all it could be an actual FieldInfo object:
https://github.com/samuelcolvin/pydantic/blob/13a5c7d676167b415080de5e6e6a74bea095b239/pydantic/main.py#L893-L896
Nor in the `ConfigError` below:
https://github.com/samuelcolvin/pydantic/blob/13a5c7d676167b415080de5e6e6a74bea095b239/pydantic/main.py#L915-L919

I wouldn't know how to update these texts though if my deductions are correct. I mean there's the easy way, but it might be a good idea to introduce a term (or use, I'm not sure it doesn't exist already) that refers to the right hand side of the assignment with field definitions (which could be the default value, an ellipsis or a FieldInfo object).

**Comment by PrettyWood** [comment]
When you write `name: Optional[str] = Field(...)` you actually set a default value to `name` with type `Optional[str]`. Personally I thought it was quite explicit that you just wrote the same in `create_model` but maybe the code or/and the public documentation could be more explicit

**Comment by MatrixScenary** [comment]
That's clear, however my problem is that _default value_ is pretty ambiguous here. It means both the right hand side of the assignment, and the _actual_ default value that the field will be initialized with. In my case, I took it granted that it means the actual default value (all the examples were instances of that) whereas it was the other way around. Docs like this further support the _wrong_ view:
https://github.com/samuelcolvin/pydantic/blob/13a5c7d676167b415080de5e6e6a74bea095b239/pydantic/main.py#L893-L896
If you look at how `<name>=<default value>` and `<name>=<FieldInfo>` are distinctly presented.

**Comment by PrettyWood** [comment]
Agreed! I put the documentation label. Feel free to suggest a PR if you want :)

**Comment by MatrixScenary** [comment]
As I'm thinking more about this, I believe only a holistic resolution would suffice, as this term is present in numerous places. Instances in which _default value_ refers to the _value-to-init-with_ should be left as such, and in the other cases, where it refers to the right-hand side value, it should be changed (to "right-hand side value (of assignment)" I guess) so it's unequivocal.

Example of the former:
https://github.com/samuelcolvin/pydantic/blob/master/docs/usage/models.md#basic-model-usage
> User here is a model with two fields id which is an integer and is required, and name which is a string and is not required (it has a **default value**).

Latter:
https://github.com/samuelcolvin/pydantic/blob/master/docs/usage/models.md#field-ordering
> As of v1.0 all fields with annotations (whether annotation-only or with a **default value**)...

I will take note when I encounter it and later work myself towards proposing the PR. Though I'm making the assumption here that everybody agrees with me on the generality of this issue.
